masked_softmax_backward

george-qi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71502

Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

george-qi committed:
initial commit for masked_softmax_backward

[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi commented:
Seems like the original implementation had some poor handling of memory causing the segfaults, so those should be fixed now

For testing, I added two:

1) OpInfo 
- Fails on one test (`TestGradients.test_fn_grad`), which I've elaborated more in the comments. 
2) Test with `gradcheck` under `test/test_nn.py`
- Fails with CUDA but succeeds with CPU

I suspect that the CPU is fine but there's some way of adding something to the test to make the gradients work, but the CUDA implementation has a bug right now

cc @cpuhrsch

george-qi commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -14948,19 +14977,36 @@ def ref_pairwise_distance(input1, input2):
             DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),
+            DecorateInfo(unittest.skip("Skipped!"), 'TestCommon', 'test_noncontiguous_samples'),
+            # torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0
+            DecorateInfo(unittest.expectedFailure, "TestGradients", "test_fn_grad", dtypes=(torch.float64,)),
```
For CPU, I'm getting a failure in `TestGradients.test_fn_grad`. For a simple case, it's failing with the following comparison: 

```
Numerical:
 tensor([[0.]], dtype=torch.float64)
Analytical:
tensor([[nan]], dtype=torch.float64)
```

From a glance, it seems to be okay, because in small cases, the non-zero gradients seem to match, and the nan's correspond to 0.

I tried adding the `gradcheck_wrapper` to use `torch.where` to convert everything with a `True` mask to 0, but that still didn't work

george-qi replied:
After this push, the CUDA implementation is completely resolved, but CPU is still running into this Numerical/Analytical difference. A bigger example of this: P476681114

george-qi commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -156,10 +156,7 @@ void host_softmax(
           scalar_t* output_data =
               output_data_base + outer_idx * outer_stride + inner_idx;
-          bool* mask_data = nullptr;
-          if (MaskedSoftMax) {
```
This was the bug

george-qi commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -156,10 +156,7 @@ void host_softmax(
-          bool* mask_data = nullptr;
-          if (MaskedSoftMax) {
-            mask_data = mask_data_base + outer_idx * outer_stride + inner_idx;
-          }
```
I believe this was causing the segfault

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

ngimel commented:
cc @zrphercule

ngimel commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -1008,5 +1024,29 @@ Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
+    const Tensor& grad,
+    const Tensor& output,
+    int64_t dim,
+    const Tensor& mask) {
```
only some of the paths in host_softmax_backward are patched to handle mask, yet you don't check here that those paths will be taken, and don't provide fallback.

george-qi replied:
This should be fixed now!

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -914,7 +920,8 @@ TORCH_IMPL_FUNC(log_softmax_backward_cuda_out) (
         "expected input and grad types to match, or input to be at::Half and grad to be at::Float");
   }
-  host_softmax_backward<LogSoftMaxBackwardEpilogue,true>(grad, output, dim, half_to_float, grad_input);
+  Tensor mask = at::ones_like(grad, ScalarType::Bool);
```
Thanks for the comment, removed this and ended up calling `dispatch_softmax_backward` directly

george-qi commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -1008,5 +1024,29 @@ Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
+    const Tensor& output,
+    int64_t dim,
+    const Tensor& mask) {
+  auto grad_ = grad.contiguous();
```
Calling `dispatch_softmax_backward`, so keeping the contiguous calls

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -783,15 +783,17 @@ Tensor host_softmax(const Tensor & input_, const int64_t dim_, const bool half_t
+  auto mask = mask_.contiguous();
   static_assert(std::is_same<acc_type<at::Half, true>, float>::value, "accscalar_t for half should be float");
   if (grad.dim() == 0) grad = grad.view(1);
+  if (mask.dim() == 0) mask = mask.view(1);
```
By broadcast, do you mean similar to how `masked_softmax_cuda` is handling the mask with `is_transformer_mask`?

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

ngimel commented:
Please add the benchmarks making sure that
a) non-masked implementation didn't regress
b) the speed-up from masked implementation
Look for benchmarks for forward softmax, @zrphercule can help you with this

ngimel commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -156,10 +156,7 @@ void host_softmax(
-          if (MaskedSoftMax) {
-            mask_data = mask_data_base + outer_idx * outer_stride + inner_idx;
-          }
+          bool* mask_data = mask_data_base + outer_idx * outer_stride + inner_idx;
```
can you please investigate why this was causing segfaults? `mask_data` is accessed only if `MaskedSoftMax` is `true`. What you have here potentially has null pointer arithmetics which is an error on llvm 12 (if softmax is not masked, `mask_data_base` should be `nullptr`, right?

george-qi replied:
Looked into this more and I think you're right - I fixed this in the newest commit and I'm not running into segfaults anymore

ngimel commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -247,14 +246,17 @@ void host_softmax_backward(
               output_data_base + outer_idx * outer_stride + inner_idx;
           const scalar_t* gradOutput_data =
               gradOutput_data_base + outer_idx * outer_stride + inner_idx;
+          bool* mask_data = mask_data_base + outer_idx * outer_stride + inner_idx;
```
and, again, you can't do that if `mask_data` is `nullptr`

ngimel commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -939,18 +939,26 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
-    // Always do masked softmax on last dim
-    int softmax_elements = input.size(input.dim() - 1);
+Tensor masked_softmax_cuda(const Tensor& input_, int64_t dim, const Tensor& mask_) {
+    // TORCH_CHECK(mask.scalar_type() == ScalarType::Bool, "Mask should be a boolean tensor");
```
clean this up?

ngimel commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -939,18 +939,26 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
+    bool is_transformer_mask = (input_.dim() == 4 && mask_.dim() == 2 && input_.size(0) == mask_.size(0) && input_.size(2) == mask_.size(1) && input_.size(3) == mask_.size(1));
+    TORCH_CHECK(mask_.sizes() == input_.sizes() || is_transformer_mask, "Mask shape should match input");
+
+    auto input = input_.dim() == 0 ? input_.view(1) : input_;
```
please add `masked_softmax` tests for 0-dim inputs (that would have caught this)

george-qi replied:
Added!

ngimel commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -939,18 +939,26 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
     // Therefore once softmax_elements > 1024, we need to fallback to vanilla masked_softmax
-    Tensor output = at::empty_like(input, input.options());
     // Fallback to a slower masked softmax solution
     if (softmax_elements > 1024 || softmax_elements * input.element_size() > 4096 || !mask.is_contiguous()) {
```
when `dim` is not the last, you also should take this fallback

george-qi replied:
ah yes, good catch! I realized this today while debugging it as well

ngimel commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -1008,5 +1016,71 @@ Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
+  grad = grad * output;
+  int64_t dim = maybe_wrap_dim(dim_, grad.dim());
+
+  TORCH_CHECK(dim >=0 && dim < grad.dim(), "dim must be non-negative and less than input dimensions");
```
you don't need this, `maybe_wrap_dim` will error out on out-of-range dim (that's why you are sending second argument to it)

ngimel commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -1008,5 +1016,71 @@ Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
+  int softmax_elements = output.size(dim);
+  int64_t batch_count = grad.numel() / softmax_elements;
+
+  if (softmax_elements > 1024) {
```
shouldn't there be condition of `softmax_elements * element_size`, like in forward softmax?

ngimel commented on test/test_nn.py:
```diff
@@ -16020,41 +16020,57 @@ def test_masked_softmax(self, device):
             self.assertEqual(pt_res, native_res, exact_dtype=True)
 
+    def test_masked_softmax_grad(self, device):
+        sizes = [(1, 1, 32), (3, 16, 310), (12, 4, 1024)]
```
please add a bigger input to test fallback

ngimel commented on test/test_nn.py:
```diff
@@ -16020,41 +16020,57 @@ def test_masked_softmax(self, device):
+            if self.device_type == "meta":
+                return
+            mask = mask.reshape(B, 1, 1, L).expand(B, num_heads, L, L).bool()
+            gradcheck(torch._masked_softmax, (input, dim, mask))
```
running gradcheck on big inputs is wasteful, instead you could compare your computed gradient to naive implementation (where you are masking results)

george-qi replied:
Thanks for pointing this out - checking the computed gradients now

ngimel commented on test/test_nn.py:
```diff
@@ -16020,41 +16020,57 @@ def test_masked_softmax(self, device):
         sizes = [(1, 1, 32), (3, 16, 310), (12, 4, 1024), (4, 2, 1200)]
         for (B, num_heads, L) in sizes:
             input = torch.randn((B, num_heads, L, L))
+            dim = input.dim() - 1
```
you should also be checking other `dim` arguments now that you've made it an input arg

george-qi replied:
Checking `0` and `input.dim()-1` now!

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -1005,8 +1008,71 @@ Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
+        );
+      });
+  }
+  return at::nan_to_num(grad_input);
```
Not entirely confident this is the right approach - I want to double check the logic here. Consider the following example from `test_masked_softmax_forward_with_nans`:

```
input
tensor([[[[-0.0883,  0.3420,  0.4112,  1.0051],
          [-0.1117, -0.5988, -0.0982, -0.3511],
          [ 0.7209, -0.2169, -1.0427,  0.2448],
          [-0.9887, -0.5196,  0.6585,  0.6406]]]], device='cuda:0',
       requires_grad=True)
mask
tensor([[[[False,  True, False,  True],
          [False,  True, False,  True],
          [False,  True, False,  True],
          [False,  True, False,  True]]]], device='cuda:0')
result
tensor([[[[nan, 1., nan, 1.],
          [nan, 1., nan, 1.],
          [nan, 1., nan, 1.],
          [nan, 1., nan, 1.]]]], device='cuda:0',
       grad_fn=<MaskedSoftmaxBackward0>)
expected
tensor([[[[nan, 1., nan, 1.],
          [nan, 1., nan, 1.],
          [nan, 1., nan, 1.],
          [nan, 1., nan, 1.]]]], device='cuda:0', grad_fn=<SoftmaxBackward0>)
```

This is all expected, but the `at::nan_to_num` is necessary in order to match the gradient:
```
input_ref.grad
tensor([[[[0., 0., 0., 0.],
          [0., 0., 0., 0.],
          [0., 0., 0., 0.],
          [0., 0., 0., 0.]]]], device='cuda:0')
```
I checked the logic and even compared all the values being set in the backwards gradient from `Softmax.cpp` (where I also applied the same `nan_to_num` function, and they were identical. I'm wondering if the structured kernel takes care of this somehow by automatically converting nan's in the backwards gradient to 0?

albanD replied:
I don't think we ever want to override nans tbh.
In the test, if you generate nan value, well that's what you get in the end.

In the masked semantic, if these values are masked anyway, we don't care what they are and should remain whatever (nan in this case?).

I would also stress again that finite difference and autograd in general only works for differentiable ops. If you're doing non-differentiable stuff, you should not try to match these together.

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -465,23 +479,62 @@ Tensor log_softmax(const Tensor& self, Dimname dim, optional<ScalarType> dtype)
+            false /* LogSoftMax */,
+            true /* MaskedSoftmax */>(grad_input, grad, output, dim, mask.data_ptr<bool>());
+      });
+  return at::nan_to_num_(grad_input);
```
Not entirely confident this is the right approach - I want to double check the logic here. Consider the following example from `test_masked_softmax_forward_with_nans`:

```
input
tensor([[-0.0883,  0.3420,  0.4112,  1.0051,  0.6003],
        [-0.8769,  0.9649, -0.1926,  0.1148, -0.1952],
        [-0.5215,  0.8718, -0.5610,  0.1900, -0.1660],
        [ 0.9107, -0.3660,  0.0853,  1.0006, -0.0962]], requires_grad=True)
mask
tensor([[False,  True, False,  True, False],
        [False,  True, False,  True, False],
        [False,  True, False,  True, False],
        [False,  True, False,  True, False]])
result
tensor([[   nan, 0.1978,    nan, 0.3510,    nan],
        [   nan, 0.3688,    nan, 0.1441,    nan],
        [   nan, 0.3360,    nan, 0.1554,    nan],
        [   nan, 0.0974,    nan, 0.3495,    nan]], grad_fn=<MaskedSoftmaxBackward0>)
expected
tensor([[   nan, 0.1978,    nan, 0.3510,    nan],
        [   nan, 0.3688,    nan, 0.1441,    nan],
        [   nan, 0.3360,    nan, 0.1554,    nan],
        [   nan, 0.0974,    nan, 0.3495,    nan]], grad_fn=<SoftmaxBackward0>)
```

Note two things:

1) Getting these `nan`s in  requires dividing by zero to achieve (e.g. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L200). The test works fine locally, but then in the test suite, it complains of dividing by zero 

2) The `at::nan_to_num` is necessary in order to match the gradient (otherwise, the 0's are replaced by `nan`s:
```
input_ref.grad
tensor([[ 0.0000, -0.1066,  0.0000, -0.0283,  0.0000],
        [ 0.0000,  0.2753,  0.0000, -0.0099,  0.0000],
        [ 0.0000, -0.0511,  0.0000,  0.1728,  0.0000],
        [ 0.0000, -0.1175,  0.0000, -0.1346,  0.0000]])
```
I checked the logic and even compared all the values being set in the backwards gradient from `Softmax.cpp` (where I also applied the same `nan_to_num` function), and they were identical. I'm wondering if the structured kernel takes care of this somehow by automatically converting nan's in the backwards gradient to 0?

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"




[ghstack-poisoned]

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34037177).

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

george-qi commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -939,34 +939,39 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
+        Tensor mask_not = mask.logical_not();
+        output = at::softmax(input.masked_fill(mask_not, -std::numeric_limits<scalar_t>::infinity()), dim);
+        // masked_softmax should return 0 in the output even when falling back to vanilla softmax
+        output = output.masked_fill(mask_not, 0);
```
In the fallback functions, need to `masked_fill` to 0 in case there are `nan`s that arise. Could also use `nan_to_mean`, but this seems more correct since it will only affect those in the mask in case there are other `nan`s

george-qi commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -1005,8 +1010,72 @@ Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
+          dim,
+          grad_input.scalar_type()
+        );
+        grad_input = grad_input.masked_fill(mask_not, 0);
```
Similar argument as in the forward function!

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

zrphercule commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -203,7 +203,10 @@ void host_softmax(
             // LogSoftMax and MaskedSoftMax should not both be true
-            if (LogSoftMax) {
+            if (MaskedSoftMax && !mask_data[d * dim_stride]) {
+              output_data[d * dim_stride] = 0;
```
This is not needed, since when (MaskedSoftMax && !mask_data[d * dim_stride]) is true, output_data[...] is always 0, this is assured at line 186 and 192.

george-qi replied:
Yeah you're right - this was originally used to guarantee that if `tmpsum=0` -- i.e. if the entire row is masked -- then we would still get 0's. But we've updated the behavior to say that it *may* return all 0's, so I'll delete this

zrphercule commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -465,23 +482,62 @@ Tensor log_softmax(const Tensor& self, Dimname dim, optional<ScalarType> dtype)
 }
 
-Tensor masked_softmax_cpu(const Tensor& input, const Tensor& mask) {
+Tensor masked_softmax_cpu(const Tensor& input, int64_t dim, const Tensor& mask) {
```
Could you please make this dim optional, and set the by-default to be input.dim()-1?

george-qi replied:
Done!

zrphercule commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -465,23 +482,62 @@ Tensor log_softmax(const Tensor& self, Dimname dim, optional<ScalarType> dtype)
+      "Mask should be a boolean tensor");
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+  auto grad = grad_.contiguous();
```
contiguous() sometimes takes a lot of time. Please make sure it wont take too much time when the tensor is contiguous already, or else you should check if it is contiguous first.

george-qi replied:
I don't think this should be an issue - https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorProperties.cpp#L62-L64 indicates that if it's already contiguous, it will just return itself

ngimel replied:
`.contiguous()` is very fast when tensor is contiguous already.

zrphercule commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -939,34 +939,37 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
+
+  int softmax_elements = input.size(dim);
+  // Persistent softmax only support softmax_elements <= 1024,
+  // Therefore once softmax_elements > 1024, we need to fallback to vanilla masked_softmax
```
re-write the comments here

zrphercule commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -1005,8 +1008,71 @@ Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
+  return output;
+}
+
+Tensor masked_softmax_backward_cuda(
```
Do you need to check if the mask_ is transformer layout in this case?

george-qi replied:
As discussed offline, will resolve in a follow-up diff!

zrphercule reviewed (changes requested):
Overall looks good, has some questions & small things to fix, please ping me once they are all checked. Thanks!

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34037177).

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

cpuhrsch commented:
@george-qi - looks like the errors are real

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

cpuhrsch reviewed (approved):
@pytorchbot merge this

george-qi committed:
Update on "masked_softmax_backward"


Differential Revision: [D34037177](https://our.internmc.facebook.com/intern/diff/D34037177)

[ghstack-poisoned]

george-qi commented:
@pytorchbot merge this
